executor/insert.go

```diff
@@ -359,7 +359,7 @@ func (e *InsertExec) initEvalBuffer4Dup() {
 		evalBufferTypes = append(evalBufferTypes, &col.FieldType)
 	}
 	if extraLen > 0 {
-		evalBufferTypes = append(evalBufferTypes, e.SelectExec.base().retFieldTypes[numWritableCols:]...)
+		evalBufferTypes = append(evalBufferTypes, e.SelectExec.base().retFieldTypes[0:]...)
```

Comment:
1. `0` is not reasonable here. IMHO, we should use `e.rowLen`
2. We also have some problem in line 371, you can reproduce by using the following case.

```
drop table if exists t1;
drop table if exists t2;
create table t1(c1 int, c2 int, c3 varchar(100), unique key(c3));
create table t2(c1 int);
insert into t2 values(1);
insert into t1(c3) select "abc" from t2 on duplicate key update t1.c3 = t2.c1;
```

Reply:
yes, you're right, but it shouldn't be `e.rowlen` either.
I will think about how to solve this problem.

Reply:
Excuse me, but do we have any update yet?